netket/optimizer/sr/sr_onthefly_logic.py

```diff
@@ -91,16 +40,32 @@ def O_vjp(x, params, v, forward_fn):
     return jax.tree_map(sum_inplace, res)  # allreduce w/ MPI.SUM
 
 
-def O_mean(samples, params, forward_fn):
+def O_mean(samples, params, forward_fn, holomorphic=True):
     r"""
     compute \langle O \rangle
     i.e. the mean of the rows of the jacobian of forward_fn
     """
 
     # determine the output type of the forward pass
     dtype = jax.eval_shape(forward_fn, params, samples).dtype
-    v = jnp.ones(samples.shape[0], dtype=dtype) * (1.0 / (samples.shape[0] * n_nodes))
-    return O_vjp(samples, params, v, forward_fn)
+    w = jnp.ones(samples.shape[0], dtype=dtype) * (1.0 / (samples.shape[0] * n_nodes))
+
+    homogeneous = nkjax.tree_ishomogeneous(params)
+    real_params = not nkjax.tree_leaf_iscomplex(params)
+    real_out = not nkjax.is_complex(jax.eval_shape(forward_fn, params, samples))
+
+    if homogeneous and (real_params or holomorphic):
+        if real_params and not real_out:
```

Comment:
i don't understand what changes between nkjax.vjp and O_vjp

Reply:
its calling https://github.com/netket/netket/blob/8a0c4658879a01c3738c393fdb789f9d31d931d7/netket/jax/_vjp.py#L96

Reply:
and what does O_vjp does that nkjax.vjp does not do?

Reply:
minus the allreduce